### PR TITLE
Energy metering for persistent memory usage

### PR DESCRIPTION
**Pull request by coolreader18**
# Description of Changes

As on the tin. I think this is the idea, right? Or am I missing another part of the database that stores user data in-memory.

# Expected complexity level and risk

2

### CODE REVIEW

File: crates/core/src/database_instance_context_controller.rs [hunk 1]
```diff
@@ -144,13 +146,14 @@ impl DatabaseInstanceContextController {
         }
     }
 
-    pub fn start_disk_monitor(&self) {
-        tokio::spawn(self.clone().disk_monitor());
+    fn start_usage_monitors(self: &Arc<Self>) {
+        tokio::spawn(Self::disk_monitor(Arc::downgrade(self)));
+        tokio::spawn(Self::memory_monitor(Arc::downgrade(self)));
```
Comment (joshua-spacetime) [inline crates/core/src/database_instance_context_controller.rs:151]: The downgrade is not strictly necessary right?

File: crates/core/src/database_instance_context_controller.rs [hunk 2]
```diff
@@ -160,14 +163,43 @@ impl DatabaseInstanceContextController {
...
                         .record_disk_usage(&db.database, db.database_instance_id, disk_usage.sum(), dt);
                     *prev_disk_usage = disk_usage;
                 }
             }
+
+            prev_tick = tick;
+        }
+    }
+
+    const MEM_METERING_INTERVAL: Duration = Duration::from_secs(100);
```
Comment (joshua-spacetime) [inline crates/core/src/database_instance_context_controller.rs:182]: Why not use the same duration as disk metering?

File: crates/sats/src/algebraic_value.rs [hunk 3]
```diff
@@ -110,6 +110,32 @@ pub enum AlgebraicValue {
...
+            | AlgebraicValue::I16(_)
+            | AlgebraicValue::U16(_)
+            | AlgebraicValue::I32(_)
+            | AlgebraicValue::U32(_)
+            | AlgebraicValue::I64(_)
+            | AlgebraicValue::U64(_)
+            | AlgebraicValue::I128(_)
+            | AlgebraicValue::U128(_)
+            | AlgebraicValue::F32(_)
+            | AlgebraicValue::F64(_) => 0,
```
Comment (joshua-spacetime) [inline crates/sats/src/algebraic_value.rs:134]: Is it true that AlgebraicValues are always transient?

File: crates/sats/src/array_value.rs [hunk 4]
```diff
@@ -47,6 +49,32 @@ pub enum ArrayValue {
...
+            Self::I16(x) => x.get_heap_size(),
+            Self::U16(x) => x.get_heap_size(),
+            Self::I32(x) => x.get_heap_size(),
+            Self::U32(x) => x.get_heap_size(),
+            Self::I64(x) => x.get_heap_size(),
+            Self::U64(x) => x.get_heap_size(),
+            Self::I128(x) => x.get_heap_size(),
+            Self::U128(x) => x.get_heap_size(),
+            Self::F32(x) => x.capacity() * 4,
+            Self::F64(x) => x.capacity() * 8,
```
Comment (joshua-spacetime) [inline crates/sats/src/array_value.rs:70]: Why are these different?

File: crates/sats/src/data_key.rs [hunk 5]
```diff
@@ -16,6 +16,9 @@ pub enum DataKey {
     Hash(Hash),
 }
 
+#[cfg(feature = "get-size")]
+impl get_size::GetSize for DataKey {}
```
Comment (joshua-spacetime) [inline crates/sats/src/data_key.rs:20]: Why can't we derive here?

File: crates/standalone/src/energy_monitor.rs [hunk 6]
```diff
@@ -37,7 +37,12 @@ impl EnergyMonitor for StandaloneEnergyMonitor {
...
 
     fn record_disk_usage(&self, database: &Database, _instance_id: u64, disk_usage: u64, period: Duration) {
-        let amount = EnergyQuanta::from_disk_usage(disk_usage, period);
+        let amount = EnergyQuanta::from_data_storage(disk_usage, period);
+        self.withdraw_energy(database.identity, amount)
+    }
+
+    fn record_memory_usage(&self, database: &Database, _instance_id: u64, mem_usage: u64, period: Duration) {
+        let amount = EnergyQuanta::from_data_storage(mem_usage, period);
         self.withdraw_energy(database.identity, amount)
```
Comment (joshua-spacetime) [inline crates/standalone/src/energy_monitor.rs:46]: Note, memory consumes more energy than disk. I believe Tyler suggested a 100X scale factor over disk.

Also can you add a new parameter to `withdraw_energy` to distinguish among the different dimensions?

Comment (cloutiertyler) [reply crates/standalone/src/energy_monitor.rs:46]: I confirm that.

File: crates/core/src/database_instance_context_controller.rs (hunk 2 above)
Comment (coolreader18) [reply crates/core/src/database_instance_context_controller.rs:182]: lmao, I thought Tyler said that the interval should be 100, not that the energy multiplier should be 100.

File: crates/core/src/database_instance_context_controller.rs (hunk 1 above)
Comment (coolreader18) [reply crates/core/src/database_instance_context_controller.rs]: It makes it so the task doesn't continue on forever, even when no one else is holding onto the controller.

File: crates/standalone/src/energy_monitor.rs (hunk 6 above)
Comment (coolreader18) [reply crates/standalone/src/energy_monitor.rs]: Fixed the scale factor. I don't know if withdraw_energy needs a dimension param, cause it's just for standalone. Where would the dimensions be stored?

**Comment by bfops** [comment]
@coolreader18 I just want to confirm for PR labeling purposes - this change is backwards-compatible right?

**Comment by coolreader18** [comment]
@bfops yes, it is.

**Comment by coolreader18** [comment]
Oh, from the user side at least - private will need an update (but I have it ready already)

**Comment by coolreader18** [comment]
The hashbrown update is because 0.15 added an `allocation_size()` method, which is obviously very helpful. It also changes the default hasher, though, and so all the `HashCollectionExt` imports are because we can no longer use `HashMap::new()`.

**Review by joshua-spacetime — COMMENTED** [review]
Thanks @coolreader18! Just a few typos, but otherwise looks good.

You've accounted for nearly all mem usage :)

I think it would have been sufficient to just track the pages, indexes, blob store, and pointer map. That probably could have reduced a lot of the boilerplate. But this is very nice nonetheless.

File: crates/table/src/memory_usage.rs [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+                .map(|(k, v)| k.memory_usage() + v.memory_usage())
+                .sum::<usize>()
+    }
+}
+
+impl<K: MemoryUsage, V: MemoryUsage> MemoryUsage for std::collections::BTreeMap<K, V> {
+    fn memory_usage(&self) -> usize {
+        self.iter()
+            .map(|(k, v)| k.memory_usage() + v.memory_usage())
+            .sum::<usize>()
```
[the remark this answers is not in the document]
Comment (coolreader18) [reply crates/table/src/memory_usage.rs:110]: Yeah, I should put a comment here that this is best-effort.

File: crates/table/src/blob_store.rs [hunk 8]
```diff
@@ -24,6 +26,8 @@ pub struct BlobHash {
     pub data: [u8; BlobHash::SIZE],
 }
 
+impl MemoryUsage for BlobHash {}
```
[the remark this answers is not in the document]
Comment (coolreader18) [reply crates/table/src/blob_store.rs:29]: `memory_usage()` is only heap memory usage - the 32 is in the `mem::size_of::<BlobHash>()`, so it doesn't need to count it in from the function.

**Review by joshua-spacetime — APPROVED** [review]

**Pull request merged**